src/ctap2/preflight.rs

```diff
@@ -0,0 +1,211 @@
+use super::client_data::ClientDataHash;
+use super::commands::get_assertion::{GetAssertion, GetAssertionOptions};
+use super::commands::{CommandError, PinUvAuthCommand, RequestCtap1, Retryable, StatusCode};
+use crate::authenticatorservice::GetAssertionExtensions;
+use crate::consts::{PARAMETER_SIZE, U2F_AUTHENTICATE, U2F_CHECK_IS_REGISTERED};
+use crate::ctap2::server::{PublicKeyCredentialDescriptor, RelyingPartyWrapper};
+use crate::errors::AuthenticatorError;
+use crate::transport::errors::{ApduErrorStatus, HIDError};
+use crate::transport::FidoDevice;
+use crate::u2ftypes::CTAP1RequestAPDU;
+use sha2::{Digest, Sha256};
+
+/// This command is used to check which key_handle is valid for this
+/// token. This is sent before a GetAssertion command, to determine which
+/// is valid for a specific token and which key_handle GetAssertion
+/// should send to the token. Or before a MakeCredential command, to determine
+/// if this token is already registered or not.
+#[derive(Debug)]
+pub(crate) struct CheckKeyHandle<'assertion> {
+    pub(crate) key_handle: &'assertion [u8],
+    pub(crate) client_data_hash: &'assertion [u8],
+    pub(crate) rp: &'assertion RelyingPartyWrapper,
+}
+
+impl<'assertion> RequestCtap1 for CheckKeyHandle<'assertion> {
+    type Output = ();
+    type AdditionalInfo = ();
+
+    fn ctap1_format(&self) -> Result<(Vec<u8>, Self::AdditionalInfo), HIDError> {
+        let flags = U2F_CHECK_IS_REGISTERED;
+        // TODO(MS): Need to check "up" here. If up==false, set to 0x08? Or not? Spec is
+        // ambiguous
```

Comment:
Ok, yes it makes sense. I do see that we could skip the filtering-step here for `up=false`, but that would be another major refactoring to actually implement properly. And I don't really see `up=false` being used in practice, so I think its ok to send one additional request to the token in that case, that ideally could be skipped.

I'll update the comment accordingly, and actually also fix the serialization of GetAssertion, because that sets the flag to zero, in case of `up=false`, which is explicitly forbidden by the CTAP1-spec. (I honestly only copied the existing code here, and didn't double-check everything).